linkerd/router/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+#![deny(rust_2018_idioms, clippy::disallowed_methods, clippy::disallowed_types)]
+#![forbid(unsafe_code)]
+
+use ahash::AHashMap;
+use futures::{future, prelude::*};
+use linkerd_error::Error;
+use linkerd_stack::{layer, NewService, Oneshot, Service, ServiceExt};
+use parking_lot::Mutex;
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    marker::PhantomData,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tracing::{debug, error};
+
+pub trait SelectRoute<Req> {
+    type Key: Eq + Hash + Clone + Debug + Send + Sync + 'static;
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    /// Given a a request, returns the key matching this request.
+    ///
+    /// If no route matches the request, this method returns an error.
+    fn select<'r>(&self, req: &'r Req) -> Result<&Self::Key, Self::Error>;
+}
+
+/// A [`NewService`] that lazy builds stacks for each `K`-typed key.
```

Comment:
nit: technically it's the `Route` that lazily builds stacks for each `K`-typed key; this just builds `Route`s...